fhir-codegen/gradle-plugin/src/main/kotlin/com/google/fhir/codegen/primitives/EnumerationFileSpecGenerator.kt

```diff
@@ -0,0 +1,186 @@
+/*
+ * Copyright 2025 Google LLC
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *       http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package com.google.fhir.codegen.primitives
+
+import com.google.fhir.codegen.schema.sanitizeKDoc
+import com.squareup.kotlinpoet.ClassName
+import com.squareup.kotlinpoet.FileSpec
+import com.squareup.kotlinpoet.FunSpec
+import com.squareup.kotlinpoet.KModifier
+import com.squareup.kotlinpoet.ParameterSpec
+import com.squareup.kotlinpoet.ParameterizedTypeName.Companion.parameterizedBy
+import com.squareup.kotlinpoet.PropertySpec
+import com.squareup.kotlinpoet.STAR
+import com.squareup.kotlinpoet.STRING
+import com.squareup.kotlinpoet.TypeSpec
+import com.squareup.kotlinpoet.TypeVariableName
+import com.squareup.kotlinpoet.asTypeName
+import kotlin.text.substringAfterLast
+
+/**
+ * Generates [FileSpec] for a FHIR `Enumeration` type bound to a specific set of codes. It is an
+ * extension to primitive type `code` and represents a constrained code value from an enumerated
+ * list
+ */
+object EnumerationFileSpecGenerator {
+  private const val R4 = "r4"
+  private const val R4B = "r4b"
+  private const val R5 = "r5"
+
+  fun generate(packageName: String): FileSpec {
+    val version = packageName.substringAfterLast('.')
+
+    val baseClassName =
+      when (version) {
+        R4,
+        R4B -> "Element"
+        R5 -> "PrimitiveType"
+        else -> error("Enum generation is not supported for version: $version")
+      }
+    val extensionClassName = ClassName(packageName, "Extension")
+
+    // Define T with upper bound of Enum<*>
+    val enumType = ClassName("kotlin", "Enum").parameterizedBy(STAR)
+    val typeVariable = TypeVariableName("T", enumType)
+
+    val elementClassName = ClassName(packageName, "Element")
+    val toElementFunction = createToElementFunction(elementClassName)
+    val ofFunction = createOfFunction(elementClassName, typeVariable)
+    val companionObject = createCompanionObject(ofFunction)
+
+    val enumClass =
```

Comment:
i would inline this whole thing too.

you can also use `apply` on the builder and create some helper functions, such as in the model type generator. up to you.